File: lib/bindings/langs/swift/Sources/BreezSDKLiquid/Task/InvoiceRequest.swift
import UserNotifications
import Foundation

struct InvoiceRequestRequest: Codable {
    let offer: String
    let invoice_request: String
    let reply_url: String
}

struct InvoiceRequestResponse: Decodable, Encodable {
    let invoice: String
    
    init(invoice: String) {
        self.invoice = invoice
    }
}

struct InvoiceErrorResponse: Decodable, Encodable {
    let error: String
    
    init(error: String) {
        self.error = error
    }
}

class InvoiceRequestTask : ReplyableTask {
    fileprivate let TAG = "InvoiceRequestTask"
    
    init(payload: String, logger: ServiceLogger, contentHandler: ((UNNotificationContent) -> Void)? = nil, bestAttemptContent: UNMutableNotificationContent? = nil) {
        let successNotificationTitle = ResourceHelper.shared.getString(key: Constants.INVOICE_REQUEST_NOTIFICATION_TITLE, fallback: Constants.DEFAULT_INVOICE_REQUEST_NOTIFICATION_TITLE)
        let failNotificationTitle = ResourceHelper.shared.getString(key: Constants.INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE, fallback: Constants.DEFAULT_INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE)
        super.init(payload: payload, logger: logger, contentHandler: contentHandler, bestAttemptContent: bestAttemptContent, successNotificationTitle: successNotificationTitle, failNotificationTitle: failNotificationTitle)
    }
    
    override func start(liquidSDK: BindingLiquidSdk) throws {
        var request: InvoiceRequestRequest? = nil
        do {
            request = try JSONDecoder().decode(InvoiceRequestRequest.self, from: self.payload.data(using: .utf8)!)
        } catch let e {
            self.logger.log(tag: TAG, line: "failed to decode payload: \(e)", level: "ERROR")
            self.displayPushNotification(title: self.failNotificationTitle, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_REPLACEABLE)
            throw e
        }
        
        do {
            let createBolt12InvoiceRes = try liquidSDK.createBolt12Invoice(req: CreateBolt12InvoiceRequest(offer: request!.offer, invoiceRequest: request!.invoice_request))
            self.replyServer(encodable: InvoiceRequestResponse(invoice: createBolt12InvoiceRes.invoice), replyURL: request!.reply_url)
        } catch let e as PaymentError {
            self.logger.log(tag: TAG, line: "failed to process invoice request: \(e)", level: "ERROR")
            let error = e.localizedDescription
            self.notifyError(request: request, error: error)
        } catch let e {
            self.logger.log(tag: TAG, line: "failed to process invoice request: \(e)", level: "ERROR")
            self.notifyError(request: request, error: "Failed to process invoice request")
        } 
    }

    func notifyError(request: InvoiceRequestRequest?, error: String) {
        if request != nil {
            self.replyServer(encodable: InvoiceErrorResponse(error: error), replyURL: request!.reply_url)
        }
        self.displayPushNotification(title: self.failNotificationTitle, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_REPLACEABLE)
    }
}

File: lib/bindings/langs/android/lib/src/main/kotlin/breez_sdk_liquid_notification/job/InvoiceRequest.kt
package breez_sdk_liquid_notification.job

import android.content.Context
import breez_sdk_liquid.CreateBolt12InvoiceRequest
import breez_sdk_liquid.PaymentException
import breez_sdk_liquid.SdkEvent
import breez_sdk_liquid.BindingLiquidSdk
import breez_sdk_liquid_notification.Constants.DEFAULT_INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE
import breez_sdk_liquid_notification.Constants.DEFAULT_INVOICE_REQUEST_NOTIFICATION_TITLE
import breez_sdk_liquid_notification.Constants.INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE
import breez_sdk_liquid_notification.Constants.INVOICE_REQUEST_NOTIFICATION_TITLE
import breez_sdk_liquid_notification.Constants.NOTIFICATION_CHANNEL_REPLACEABLE
import breez_sdk_liquid_notification.NotificationHelper.Companion.notifyChannel
import breez_sdk_liquid_notification.ResourceHelper.Companion.getString
import breez_sdk_liquid_notification.SdkForegroundService
import breez_sdk_liquid_notification.ServiceLogger
import breez_sdk_liquid_notification.PluginConfigs
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class InvoiceRequestRequest(
    @SerialName("offer") val offer: String,
    @SerialName("invoice_request") val invoiceRequest: String,
    @SerialName("reply_url") val replyURL: String,
)

@Serializable
data class InvoiceRequestResponse(
    val invoice: String,
)

@Serializable
data class InvoiceErrorResponse(
    val error: String,
)

class InvoiceRequestJob(
    private val context: Context,
    private val fgService: SdkForegroundService,
    private val payload: String,
    private val logger: ServiceLogger,
) : Job {
    companion object {
        private const val TAG = "InvoiceRequestJob"
    }

    override fun start(liquidSDK: BindingLiquidSdk, pluginConfigs: PluginConfigs) {
        var request: InvoiceRequestRequest? = null
        try {
            val decoder = Json { ignoreUnknownKeys = true }
            request = decoder.decodeFromString(InvoiceRequestRequest.serializer(), payload)
            val createBolt12InvoiceResponse =
                liquidSDK.createBolt12Invoice(
                    CreateBolt12InvoiceRequest(request.offer, request.invoiceRequest),
                )
            val response = InvoiceRequestResponse(createBolt12InvoiceResponse.invoice)
            val success = replyServer(Json.encodeToString(response), request.replyURL)
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_REPLACEABLE,
                getString(
                    context,
                    if (success) INVOICE_REQUEST_NOTIFICATION_TITLE else INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE,
                    if (success) DEFAULT_INVOICE_REQUEST_NOTIFICATION_TITLE else DEFAULT_INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE,
                ),
            )
        } catch (e: PaymentException) {
            logger.log(TAG, "Failed to process invoice request: ${e.message}", "WARN")
            notifyError(request, e.message ?: "Failed to create invoice")
        } catch (e: Exception) {
            logger.log(TAG, "Failed to process invoice request: ${e.message}", "WARN")
            notifyError(request, "Failed to create invoice")
        }

        fgService.onFinished(this)
    }

    private fun notifyError(request: InvoiceRequestRequest?, error: String) {
        if (request != null) {
            val errorResponse = InvoiceErrorResponse(error)
            replyServer(Json.encodeToString(errorResponse), request.replyURL)
        }
        notifyChannel(
            context,
            NOTIFICATION_CHANNEL_REPLACEABLE,
            getString(
                context,
                INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE,
                DEFAULT_INVOICE_REQUEST_NOTIFICATION_FAILURE_TITLE,
            ),
        )
    }


    override fun onEvent(e: SdkEvent) {}

    override fun onShutdown() {}
}
